frontend_build/src/extract_$trs.js

```diff
@@ -0,0 +1,96 @@
+              && ((node.expression.left || {}).property || {}).name == 'exports'
+              // Does the right hand side of the assignment expression have any properties?
+              // (We don't want to both parsing it if it is an empty object)
+              && node.expression.right.properties) {
+              // Look through each of the properties in the object that is being exported.
+              node.expression.right.properties.forEach(function (property){
+                // If the property is called $trs we have hit paydirt! Some messages for us to grab!
+                if (property.key.name === '$trs') {
+                  // Grab every message in our $trs property and save it into our messages object.
+                  property.value.properties.forEach(function(message) {
+                    messages[message.key.name] = message.value.value;
+                  });
+                  // We also want to take a note of the name space these messages have been put in too!
+                } else if (property.key.name === '$trNameSpace') {
+                  messageNameSpace = property.value.value;
+                }
+              });
+            }
+          });
+          if (messageNameSpace) {
+            // Every message needs to be namespaced - don't pollute our top level!
+            Object.keys(messages).forEach(function(key) {
+              // Create a new message id from the name space and the message id joined with '.'
+              var msgId = messageNameSpace + '.' + key;
+              if (messageExport[msgId]) {
+                // Warn about duplicate ids *within* a bundle (no way to warn across).
+                logging.warn('Duplicate translation id ' + msgId + ' found in ' + module.resource)
+              } else {
+                // If all is good, save it onto our export object for the whole bundle.
+                messageExport[msgId] = messages[key];
+              }
+            });
+            // Someone defined a $trs object, but didn't namespace it - warn them about it here so they can fix their foolishness.
+          } else if (Object.keys(messages).length) {
+            logging.warn('Translatable messages have been defined in ' + module.resource + ' but no messageNameSpace was specified.');
+          }
+        }
+      });
+    });
+    // Grab the current compilation hash to version the filename.
+    self.hash = compilation.hash || '';
+    if (Object.keys(messageExport).length) {
+      // If we've got any messages to write out, write them out. Otherwise, don't bother.
+      self.writeOutput(messageExport);
+    }
+    callback();
+  });
+};
+
+extract$trs.prototype.writeOutput = function(messageExport) {
+  // Make sure the directory we are using exists.
+  mkdirp.sync(this.messageDir);
+  // Write out the data to JSON.
+  fs.writeFileSync(path.join(this.messageDir, this.messagesName + '-' + this.hash + '-messages.json'), JSON.stringify(messageExport));
```

Comment:
@rtibbles would you be able to use the kolibri version instead of the hash? Uploading CrowdIn files with different names each time will mean that translations won't be carried over from the previous file, unless we always manually trigger auto-translations through translation memory (which takes some time).

There are ways around this on CrowdIn, but if it's easy to use the Kolibri version, I prefer to do it through here.